### PR TITLE
lister-gen updates

### PR DESCRIPTION
**Pull request by ncdc**
- Remove "zz_generated." prefix from generated lister file names
- Add support for expansion interfaces
- Switch to new generated JobLister

@deads2k @liggitt @sttts @mikedanese @caesarxuchao for the lister-gen changes
@soltysh @deads2k for the informer / job controller changes

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/36080)

### CODE REVIEW

**Comment by deads2k** [comment]
rename looks fine.  Don't clientsets generate one file per resource so that you continue to get generated_expansion files when you add new types to groups that already have customizations?

**Comment by ncdc** [comment]
@deads2k clientsets generate one file per resource plus a single generated_expansion.go per GroupVersion. If you want customized expansions, you would create e.g. pod_expansion.go and that would result in the PodExpansion interface being removed from generated_expansions.go. Does that answer your question?

**Comment by deads2k** [comment]
> @deads2k clientsets generate one file per resource plus a single generated_expansion.go per GroupVersion. If you want customized expansions, you would create e.g. pod_expansion.go and that would result in the PodExpansion interface being removed from generated_expansions.go. Does that answer your question?

Yes, and that means this lgtm.

**Comment by ncdc** [comment]
I need to fix some data race issues in the tests first :)

File: pkg/controller/job/jobcontroller.go [hunk 1]
```diff
@@ -135,39 +117,26 @@ func NewJobController(podInformer cache.SharedIndexInformer, kubeClient clientse
 	return jm
 }
 
-func NewJobControllerFromClient(kubeClient clientset.Interface, resyncPeriod controller.ResyncPeriodFunc) *JobController {
```
Comment (ncdc) [inline pkg/controller/job/jobcontroller.go]: I realized after I moved this to jobcontroller_test.go that maybe someone outside of kube could potentially be using this. Is it ok to unexport/move or do you all want it to remain (it's only used by test code in kube)? @soltysh

Comment (deads2k) [inline pkg/controller/job/jobcontroller.go]: > I realized after I moved this to jobcontroller_test.go that maybe someone outside of kube could potentially be using this. Is it ok to unexport/move or do you all want it to remain (it's only used by test code in kube)? @soltysh

Deleting it is fine.  It's cleanup we needed to do and we'll refactor downstream on the rebase to handle it.

Comment (soltysh) [inline pkg/controller/job/jobcontroller.go]: Was about to say that. @deads2k added that when [switching to shared informer](https://github.com/kubernetes/kubernetes/commit/8c4e3af1), it's used only in tests, nowhere else.

**Comment by soltysh** [comment]
JobController changes LGTM, overall this LGTM.

**Comment by ncdc** [comment]
Rebased and squashed. PTAL

**Comment by deads2k** [comment]
I'd squash in bazel updates.  They're frustrating, but not very interesting.

lgtm

**Comment by ncdc** [comment]
Bazel changes folded in. @caesarxuchao this is ready for final review - thanks!

**Comment by soltysh** [comment]
LGTM as well

**Comment by ncdc** [comment]
Working on fixing golint

**Comment by ncdc** [comment]
trusty permission denied issue

@k8s-bot node e2e test this

**Comment by ncdc** [comment]
```
Timed out waiting for cluster initialization. Cluster API may not be available
```

@k8s-bot cvm gke e2e test this

**Comment by ncdc** [comment]
@k8s-bot gci gke e2e test this

**Comment by ncdc** [comment]
Rebased

**Comment by jbeda** [comment]
Please don't rename away form `zz_generated.*`. This will create real problems as we work toward a dockerized build process.  We need a way to figure out what is generated so we don't accidentally clobber *real* sources.  If there were a way to have the output be separated into unique directories in go it would make this easier.  That can't happen so we are stuck with a naming hack.  See #38712.

**Pull request merged**